src/mmfatl.c

```diff
@@ -0,0 +1,1591 @@
+/*****************************************************************************/
+/*        Copyright (C) 2022  Wolf Lammen                                    */
+/*            License terms:  GNU General Public License                     */
+/*****************************************************************************/
+/*34567890123456 (79-character line to adjust editor window) 2345678901234567*/
+
+/*!
+ * \file mmfatl.c a self-contained set of text printing routines using
+ * dedicated pre-allocated memory, designed to likely run even under difficult
+ * conditions (corrupt state, out of memory).
+ */
+
+#include "mmfatl.h"
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdarg.h>
+#include <string.h>
+
+#ifdef TEST_MMFATL
+    // We intercept malloc/free to test proper
+    // freeing after allocation of memory
+    static void* MALLOC(size_t size);
+    static void FREE(void* ptr);
+#else
+#   define FREE free
+#   define MALLOC malloc
+#endif
+
+/*!
+ * C has no boolean type, instead an int is used.  To ease code reading we
+ * encode one boolean value here.
+ * int cond = FALSE; if (cond) { } skips the code block.
+ */
+#define FALSE 0
+/*!
+ * C has no boolean type, instead an int is used.  To ease code reading we
+ * encode one boolean value here.
+ * int cond = TRUE; if (cond) { } branches into the code block.
+ */
+#define TRUE 1
+
+/*! terminating character of a text string never used elsewhere */
+#define NUL '\x00'
+
+/*!
+ * utility function:
+ * adds two non-zero values and tests for overflow.  Zeros are indicative of a
+ * previous overflow and are propagated.
+ * \param x addend, 0 indicates it is the result of a previous overflow
+ * \param y addend, 0 indicates it is the result of a previous overflow
+ * \returns the sum of \p x and \p y, 0 indicates an overflow.
+ */
+static size_t addCheckOverflow(size_t x, size_t y)
+{
+    size_t result = x + y;
+    return x == 0 || result <= x? 0 : result;
+}
+
+/*!
+ * utility function:
+ * In case of truncation due to a full buffer the last UTF-8 multi-byte
+ * sequence may not be transferred completely.  Trace back to the last
+ * character of a completely transferred sequence.  ASCII encoding is a single byte
+ * subset encoding of UTF-8 that always yields the last transferred byte.
+ * \param begin [not null] a pointer to the buffer start.
+ * \param last [not null] a pointer to the last transferred byte in the buffer.
+ * \returns a pointer to the last valid byte, or NULL if it does not exist.
+ *   The latter can happen only if either your buffer is less than 4 bytes
+ *   long, or if it contains invalid UTF-8 encoding.
+ */
+
+static char* findLastUTF8Byte(char const* begin, char* end)
```

Comment:
We don't support UTF8 in the first place, what is this?

Reply:
Future extension.  I'd like Unicode support all over Metamath in the end.  You need to start somewhere.